source/lib/pwm.c

```diff
@@ -0,0 +1,183 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "stddef.h"
+#include "lib/ticker.h"
+#include "nrf_gpio.h"
+#include "py/runtime.h"
+#include "py/gc.h"
+
+#define PWM_TICKER_INDEX 2
+
+typedef struct _pwm_event {
+    uint32_t pins;
+    struct _pwm_event *next;
+    uint16_t timestamp;
+    bool set;
+} pwm_event;
+
+// Default period of 20ms
+#define DEFAULT_PERIOD ((20*1000)/MICROSECONDS_PER_TICK)
+
+static pwm_event all_off = {
+    .pins = 0,
+    .next = &all_off,
+    .timestamp = DEFAULT_PERIOD,
+    .set = false
+};
+
+int32_t pwm_get_period_ticks(void) {
+    return all_off.timestamp;
+}
+
+#if 0
+void pwm_dump_state(void) {
+    pwm_event *event = &all_off;
+    do {
+        printf("Event. pins: %d, timestamp: %d, set: %d\n", event->pins, event->timestamp, event->set);
+        event = event->next;
+    } while (event != &all_off);
+}
+#endif
+
+int32_t pwm_callback(void) {
+    pwm_event *event = (pwm_event *)MP_STATE_PORT(pwm_next_event);
+    if (event->set) {
+        nrf_gpio_pins_set(event->pins);
+    } else {
+        nrf_gpio_pins_clear(event->pins);
+    }
+    int32_t tnow = event->timestamp;
+    MP_STATE_PORT(pwm_next_event) = event->next;
+    int32_t tnext = event->next->timestamp;
+    int32_t tdiff = tnext - tnow;
+    if (tdiff <= 0) {
+        tdiff += pwm_get_period_ticks();
+    }
+    return tdiff;
+}
+
+void pwm_start(void) {
+    MP_STATE_PORT(pwm_next_event) = &all_off;
+    set_ticker_callback(PWM_TICKER_INDEX, pwm_callback, 120);
+}
+
+void pwm_stop(void) {
+    clear_ticker_callback(PWM_TICKER_INDEX);
+}
+
+static void pwm_turn_off_pin(int32_t pin) {
+    pwm_event *previous = &all_off;
+    pwm_event *event = all_off.next;
+    while (event != &all_off) {
+        if (event->pins & (1UL << pin)) {
+            goto found;
+        }
+        previous = event;
+        event = event->next;
+    }
+    return;
+found:
+    event->pins &= ~(1UL << pin);
+    if (event->pins == 0) {
+        // GC will clean up this event when it is no longer reachable
+        previous->next = event->next;
+    }
+    nrf_gpio_pin_clear(pin);
+    all_off.pins  &= ~(1UL << pin);
+}
+
+static void pwm_insert_pin_on_event(int8_t pin, uint16_t timestamp) {
+    nrf_gpio_cfg_output(pin);
+    pwm_event *previous = &all_off;
+    pwm_event *event = all_off.next;
+    all_off.pins |= (1UL<<pin);
+    while (event != &all_off) {
+        if (event->timestamp >= timestamp) {
+            break;
+        }
+        previous = event;
+        event = event->next;
+    }
+    if (event->timestamp == timestamp) {
+        event->pins |= (1UL<<pin);
+    } else {
+        event = gc_alloc(sizeof(pwm_event), false);
```

Comment:
It could, but then we would be using RAM even if PWM is unused.
This way we don't need to pre-determine how many pins can be supported.